### ISSUE TITLE
tf/NPO: Implement SOA entropy regularization methods and choose a sane default

### ISSUE THREAD

**Issue by ryanjulian** [rlworkgroup/garage#283]
The difference between "maximum entropy" algorithms vs "entropy regularized" algorithms is that MaxEnt algorithms augment the individual time steps rewards with the per-time policy entropy at that step, while entropy regularized algorithms augment the overall policy gradient surrogate objective with regularizer (e.g. bonus) for the average policy entropy.

Generally the maximum-entropy formulation should produce lower variance gradient estimates and is therefore preferable (though it can be harder to tune).

We can provide the user with both options if we'd like (e.g. for parity with baselines), but we should definitely support MaxEnt at least.

Here is the diff to make the current NPO MaxEnt instead of entropy-regularized:

The current NPO does neither of these. It add the average entropy for the entire batch to each rewards step:
```python
        with tf.name_scope("policy_entropy"):
            policy_dist_info_flat = self.policy.dist_info_sym(
                i.flat.obs_var,
                i.flat.policy_state_info_vars,
                name="policy_dist_info_flat")

            policy_entropy_flat = self.policy._dist.entropy_sym(
                policy_dist_info_flat)
            policy_entropy = tf.reshape(policy_entropy_flat,
                                        [-1, self.max_path_length])

            # This prevents entropy from becoming negative for small policy std
            if self._use_softplus_entropy:
                policy_entropy = tf.nn.softplus(policy_entropy)

            policy_entropy = tf.reduce_mean(policy_entropy * i.valid_var)

           with tf.name_scope("augmented_rewards"):
            rewards = i.reward_var + (self.policy_ent_coeff * policy_entropy)
```

This would be MaxEnt
```python
        with tf.name_scope("policy_entropy"):
            policy_dist_info_flat = self.policy.dist_info_sym(
                i.flat.obs_var,
                i.flat.policy_state_info_vars,
                name="policy_dist_info_flat")

            policy_entropy_flat = self.policy._dist.entropy_sym(
                policy_dist_info_flat)
            policy_entropy = tf.reshape(policy_entropy_flat,
                                        [-1, self.max_path_length])

            # This prevents entropy from becoming negative for small policy std
            if self._use_softplus_entropy:
                policy_entropy = tf.nn.softplus(policy_entropy)

           with tf.name_scope("augmented_rewards"):
            rewards = i.reward_var + (self.policy_ent_coeff * policy_entropy)
```
This would be entropy-regularized
```python
        with tf.name_scope("policy_entropy"):
            policy_dist_info_flat = self.policy.dist_info_sym(
                i.flat.obs_var,
                i.flat.policy_state_info_vars,
                name="policy_dist_info_flat")

            policy_entropy_flat = self.policy._dist.entropy_sym(
                policy_dist_info_flat)
            policy_entropy = tf.reshape(policy_entropy_flat,
                                        [-1, self.max_path_length])

            # This prevents entropy from becoming negative for small policy std
            if self._use_softplus_entropy:
                policy_entropy = tf.nn.softplus(policy_entropy)

            policy_entropy = tf.reduce_mean(policy_entropy * i.valid_var)

                # Maximize E[surrogate objective] by minimizing
                # -E_t[surrogate objective]
                surr_loss = -tf.reduce_mean(surr_obj) + (self.policy_ent_coeff * policy_entropy)
```

**Comment by ryanjulian** [comment]
I have learned there are actually several possibilities here. Let's upgrade the scope of this bug to implement all of them and pick a sane default.

See https://arxiv.org/abs/1805.00909 for a decent overview.

The configuration possibilities are:
* Adding entropy to the reward for each time step or adding mean entropy to the surrogate objective (the former is "entropy-regularized" and the latter is "max entropy"). Entropy regularization should produce a lower-variance policy gradient

![image](https://user-images.githubusercontent.com/1376219/44303420-747f7f00-a2f5-11e8-8057-3ab9bd6e1a82.png)

```python
# Entropy regularized
policy_entropy = policy.entropy(actions)
augmented_reward = r + (self._policy_ent_coeff * policy_entropy)

# MaxEnt
policy_entropy = policy.entropy(actions)
# ...
surr_loss = -tf.reduce_mean(lr * adv) - (self._policy_ent_coeff * tf.reduce_mean(policy_entropy))
```

* Calculating entropy using the formal definition, or estimating is as the negative log likelihood of the action
![image](https://user-images.githubusercontent.com/1376219/44303428-b4defd00-a2f5-11e8-8153-9a441a6ffcc9.png)
```python
# regular
policy_entropy = policy.entropy(actions)
# estimator
policy_entropy = -policy.logli(action)
```

* Allowing the gradients to flow through the entropy term or stopping them at the entropy term so that the policy cannot backpropagate changes to standard deviation directly. This forces the optimizer to increase entropy indirectly by choosing high-entropy actions, rather than by increasing the standard deviation of the policy.
```python
# normal
policy_entropy = calculate_entropy()

if self._stop_entropy_gradients:
    policy_entropy = tf.stop_gradient(policy_entropy)
```

* Applying softplus regularization to the entropy term or not, to prevent it from going very negative for small-std actions (already implemented)

**Comment by CatherineSue** [comment]
finished in #632
